lib/omise/ApiRequestor.php

```diff
@@ -0,0 +1,222 @@
+<?php
+namespace Omise;
+
+use Omise\Http\Response\Handler as ResponseHandler;
+use Exception;
+
+class ApiRequestor
+{
+    /**
+     * @var string
+     */
+    const OMISE_PHP_LIB_VERSION = '3.0.0-dev';
+    const OMISE_API_URL         = 'https://api.omise.co/';
+    const OMISE_VAULT_URL       = 'https://vault.omise.co/';
+
+    /**
+     * Request methods
+     * @var string
+     */
+    const REQUEST_GET    = 'GET';
+    const REQUEST_POST   = 'POST';
+    const REQUEST_PATCH  = 'PATCH';
+    const REQUEST_DELETE = 'DELETE';
+
+    /**
+     * Timeout setting
+     * @var int
+     */
+    private $OMISE_CONNECTTIMEOUT = 30;
+    private $OMISE_TIMEOUT        = 60;
+
+    /**
+     * @param string $url
+     * @param string $key
+     * @param array  $params
+     */
+    public function get($url, $key, $params = null)
+    {
+        return $this->request($url, self::REQUEST_GET, $key, $params);
+    }
+
+    /**
+     * @param string $url
+     * @param string $key
+     * @param array  $params
+     */
+    public function post($url, $key, $params = null)
+    {
+        return $this->request($url, self::REQUEST_POST, $key, $params);
+    }
+
+    /**
+     * @param string $url
+     * @param string $key
+     * @param array  $params
+     */
+    public function patch($url, $key, $params = null)
+    {
+        return $this->request($url, self::REQUEST_PATCH, $key, $params);
+    }
+
+    /**
+     * @param string $url
+     * @param string $key
+     * @param array  $params
+     */
+    public function delete($url, $key, $params = null)
+    {
+        return $this->request($url, self::REQUEST_DELETE, $key, $params);
+    }
```

Comment:
Sounds tempted. Let me think a bit.

Reply:
@jonrandy Is there any benefit in grouping them into one `__call` method rather than having 4 methods separately? (besides of reducing line of code).

Reply:
Easier to add new similar methods if any come along, and to remove old ones I guess. These scenarios seem unlikely. My main concern is it just feels fundamentally wrong to write essentially the same code 4 times

Reply:
Uhmm.. can't see much the different on `Easier to add new similar methods if any come along, and to remove old ones I guess.` (maybe can give me some code sample..). 

As having them separately is more easy to maintain and test
(I'm thinking of a case where we want to test 4 get, post, patch, and delete. By having it in one method is hard to identify problem as well as having 4 of them in one method will make the method itself becomes unnecessary fat)

Also having 4 methods separately is more friendly with developers as it will be documented and listed up as a `class' method` on  most of text-editors they use. (also can be generated in case if someone use those PHPDocBlock to generate a document). 

Note:
These 4 are an alias methods that just to make code looks a bit more easier to read.
Just from
```php
protected static function g_retrieve($clazz, $url, $publickey = null, $secretkey = null)
{
    $resource = call_user_func(array($clazz, 'getInstance'), $clazz, $publickey, $secretkey);
    $result   = $resource->apiRequestor->get($url, $resource->getResourceKey());
    $resource->refresh($result);

    return $resource;
}
```
can become 👇 without these alias methods.
```php
protected static function g_retrieve($clazz, $url, $publickey = null, $secretkey = null)
{
    $resource = call_user_func(array($clazz, 'getInstance'), $clazz, $publickey, $secretkey);
    $result   = $resource->apiRequestor->request($url, Omise\ApiRequestor::GET, $resource->getResourceKey());
    $resource->refresh($result);

    return $resource;
}
```

So I more likely want to go with the 4 methods separately or maybe better to remove all out (than having it in `__call`). (but then, it's a bit troublesome when it comes to using it through `request()` method.

Reply:
I was thinking along these lines (untested, probably bugged code):
```php
    const REQUEST_GET    = 'GET';
    const REQUEST_POST   = 'POST';
    const REQUEST_PATCH  = 'PATCH';
    const REQUEST_DELETE = 'DELETE';

    const REQUEST_METHODS = [REQUEST_GET, REQUEST_POST, REQUEST_PATCH, REQUEST_DELETE];

    /**
     * Timeout setting
     * @var int
     */
    private $OMISE_CONNECTTIMEOUT = 30;
    private $OMISE_TIMEOUT        = 60;

    /**
     * @param string $url
     * @param string $key
     * @param array  $params
     */
    public function __call($name, $args)
    {
        $requestMethodName = strtoupper($name);
        if (array_key_exists($requestMethodName, self::REQUEST_METHODS)) {
            return $this->request($args[0], $requestMethodName, $args[1], count($args) > 2 ? $args[2] : null);
        }
    }
```

Reply:
Looks nice 👍 
Let me try this.